### PR TITLE
Unifies AutoPlugin and RootPlugin

### PR DESCRIPTION
**Pull request by eed3si9n**
See #1203.

### CODE REVIEW

**Comment by huntc** [comment]
Cool. So that I understand then, and given the following:

untitled
- `SbtWebPlugin` requires = empty, trigger = noTrigger (user must add the plugin to the project)
- `SbtJsEnginePlugin` requires = `SbtWebPlugin`, trigger = `SbtWebPlugin`
- `SbtJsTaskPlugin` requires = `SbtJsEnginePlugin`, trigger = `SbtWebPlugin`
- `SbtLessPlugin` = requires = `SbtJsTaskPlugin`, trigger = `SbtWebPlugin`

Would that be correct?

**Comment by eed3si9n** [comment]
@huntc `trigger` values can either be `noTrigger` or `allRequirements` at the moment. This could extend into arbitrary plugins like what you're saying but for now keeping it simple.

**Comment by jroper** [comment]
I think PlayPlugin requires SbtJsTaskPlugin and has no trigger, while all the js plugins require SbtJsTask plugin and trigger on all requirements.

**Comment by huntc** [comment]
OK, then is this correct?
- `SbtWebPlugin` requires = empty, trigger = `noTrigger` (user must add the plugin to the project)
- `SbtJsEnginePlugin` requires = `SbtWebPlugin`, trigger = `allRequirements`
- `SbtJsTaskPlugin` requires = `SbtJsEnginePlugin`, trigger = `allRequirements`
- `SbtLessPlugin` requires = `SbtJsTaskPlugin`, trigger = `allRequirements`
- `PlayPlugin` requires = `SbtJsTaskPlugin`, trigger =  `noTrigger`

(some plugins omitted for brevity)

**Comment by eed3si9n** [comment]
@huntc According to your diagram, 
- `SbtWebdriverPlugin` requires = `SbtWebPlugin`, trigger = `allRequirements`
- `PlayPlugin` requires = `SbtWebdriverPlugin && SbtJsTaskPlugin`, trigger = `noTrigger`

File: main/src/main/scala/sbt/Plugins.scala [hunk 1]
```diff
@@ -9,13 +9,14 @@ TODO:
...
 	import Plugins._
+	import annotation.tailrec
 
 /** Marks a top-level object so that sbt will wildcard import it for .sbt files, `consoleProject`, and `set`. */
 trait AutoImport
 
 /**
 An AutoPlugin defines a group of settings and the conditions where the settings are automatically added to a build (called "activation").
-The `select` method defines the conditions and a method like `projectSettings` defines the settings to add.
+The `requires` and `trigger` method defines the conditions and a method like `projectSettings` defines the settings to add.
```
Comment (jozic) [inline main/src/main/scala/sbt/Plugins.scala]: The `requires` and `trigger` method_s_ define the conditions

**Comment by eed3si9n** [comment]
Added more a test on exclusion conflict per @jsuereth and what do you know, caught a bug.

Now the newly added check will return:

```
[error] sbt.AutoPluginException: Error determining plugins for project 'projA' in /private/var/folders/7g/l96y5q310h90xsz109qqjpmc0000gn/T/sbt_cef586da/auto-plugins-conflict/projA:
[error] Contradiction in selected plugins.  These plguins were both included and excluded: R
```

**Comment by jroper** [comment]
I like this. When can we get a new milestone of SBT?  We want to start using it ASAP.

**Comment by huntc** [comment]
I raise @jroper one Frappaccino

**Pull request closed without merging**